src/datasets.c

```diff
@@ -0,0 +1,990 @@
+
+            uint8_t hash[16];
+            if (HexToRaw((const uint8_t *)line, 32, hash, sizeof(hash)) < 0)
+                FatalError(SC_ERR_FATAL, "die: bad hash");
+
+            DataRepType rep = { .value = 0};
+            if (ParseRepLine(line+33, strlen(line)-33, &rep) < 0)
+                FatalError(SC_ERR_FATAL, "die: bad rep");
+
+            SCLogDebug("rep v:%u", rep.value);
+            if (DatasetAddwRep(set, hash, 16, &rep) < 0)
+                FatalError(SC_ERR_FATAL, "die: add failed");
+
+            cnt++;
+        }
+        else {
+            SCLogNotice("MD5 bad line len %u: %s", (uint32_t)strlen(line), line);
+        }
+    }
+
+    fclose(fp);
+    SCLogNotice("dataset: %s loaded %u records", set->name, cnt);
+    return 0;
+}
+
+static int DatasetLoadSha256(Dataset *set)
+{
+    if (strlen(set->load) == 0)
+        return 0;
+
+    SCLogNotice("dataset: %s loading from '%s'", set->name, set->load);
+
+    FILE *fp = fopen(set->load, "r");
+    if (fp == NULL) {
+        SCLogError(SC_ERR_DATASET, "fopen '%s' failed: %s",
+                set->load, strerror(errno));
+        return -1;
+    }
+
+    uint32_t cnt = 0;
+    char line[1024];
+    while (fgets(line, (int)sizeof(line), fp) != NULL) {
+        /* straight black/white list */
+        if (strlen(line) == 65) {
+            line[strlen(line) - 1] = '\0';
+            SCLogDebug("line: '%s'", line);
+
+            uint8_t hash[32];
+            if (HexToRaw((const uint8_t *)line, 64, hash, sizeof(hash)) < 0)
+                FatalError(SC_ERR_FATAL, "die");
+
+            if (DatasetAdd(set, (const uint8_t *)hash, (uint32_t)32) < 0)
+                FatalError(SC_ERR_FATAL, "die");
+            cnt++;
+
+            /* list with rep data */
+        } else if (strlen(line) > 65 && line[64] == ',') {
+            line[strlen(line) - 1] = '\0';
+            SCLogNotice("SHA-256 with REP line: '%s'", line);
+
+            uint8_t hash[32];
+            if (HexToRaw((const uint8_t *)line, 64, hash, sizeof(hash)) < 0)
+                FatalError(SC_ERR_FATAL, "die: bad hash");
+
+            DataRepType rep = { .value = 0 };
+            if (ParseRepLine(line+65, strlen(line)-65, &rep) < 0)
+                FatalError(SC_ERR_FATAL, "die: bad rep");
+            SCLogNotice("rep %u", rep.value);
+
+            if (DatasetAddwRep(set, hash, 32, &rep) < 0)
+                FatalError(SC_ERR_FATAL, "die: add failed");
+            cnt++;
+        }
+    }
+
+    fclose(fp);
+    SCLogNotice("dataset: %s loaded %u records", set->name, cnt);
+    return 0;
+}
+
+static int DatasetLoadString(Dataset *set)
+{
+    if (strlen(set->load) == 0)
+        return 0;
+
+    SCLogNotice("dataset: %s loading from '%s'", set->name, set->load);
+
+    FILE *fp = fopen(set->load, "r");
+    if (fp == NULL) {
+        SCLogError(SC_ERR_DATASET, "fopen '%s' failed: %s",
+                set->load, strerror(errno));
+        return -1;
+    }
+
```

Comment:
When the filename comes from state, I think its very likely the file may not exist yet, and should probably be created. I would expect an error on load though.

Reply:
Yeah good catch. Fixing this and adding a SV test.